crates/sui-storage/src/write_ahead_log.rs

```diff
@@ -0,0 +1,292 @@
+
+    fn abort_tx(mut self) -> SuiResult {
+        self.dead = true;
+        self.wal.abort_tx(&self.tx)
+    }
+}
+
+impl<C> Drop for DBTxGuard<'_, C>
+where
+    C: Serialize + DeserializeOwned,
+{
+    fn drop(&mut self) {
+        if !self.dead {
+            let tx = self.tx;
+            error!(digest = ?tx, "DBTxGuard dropped without explicit commit/abort");
+            self.wal.implicit_drop_tx(&tx);
+        }
+    }
+}
+
+// A WriteAheadLog implementation built on rocksdb.
+pub struct DBWriteAheadLog<C> {
+    log: DBMap<TransactionDigest, C>,
+
+    // Can't use tokio Mutex - must be accessible synchronously from drop trait.
+    recoverable_txes: Mutex<Vec<TransactionDigest>>,
+
+    // Guards the get/set in begin_tx
+    mutex_table: MutexTable<TransactionDigest>,
+}
+
+const MUTEX_TABLE_SIZE: usize = 1024;
+
+impl<C> DBWriteAheadLog<C>
+where
+    C: Serialize + DeserializeOwned,
+{
+    pub fn new<P: AsRef<Path>>(path: P) -> Self {
+        let (options, _) = default_db_options(None);
+        let db = {
+            let path = &path;
+            let db_options = Some(options.clone());
+            let opt_cfs: &[(&str, &rocksdb::Options)] = &[("tx_write_ahead_log", &options)];
+            typed_store::rocks::open_cf_opts(path, db_options, opt_cfs)
+        }
+        .expect("Cannot open DB.");
+
+        let log: DBMap<TransactionDigest, C> =
+            DBMap::reopen(&db, Some("tx_write_ahead_log")).expect("Cannot open CF.");
+
+        // Read in any digests that were left in the log, e.g. due to a crash.
+        let recoverable_txes: Vec<_> = log.iter().map(|(tx, _)| tx).collect();
+
+        Self {
+            log,
+            recoverable_txes: Mutex::new(recoverable_txes),
+            mutex_table: MutexTable::new(MUTEX_TABLE_SIZE),
+        }
+    }
+
+    fn commit_tx(&self, tx: &TransactionDigest) -> SuiResult {
+        debug!(digest = ?tx, "committing tx");
+        self.log.remove(tx).map_err(|e| e.into())
+    }
+
+    fn abort_tx(&self, tx: &TransactionDigest) -> SuiResult {
+        info!(digest = ?tx, "aborting tx");
+        self.log.remove(tx).map_err(|e| e.into())
+    }
+
+    fn implicit_drop_tx(&self, tx: &TransactionDigest) {
+        // this function should be called very rarely so contention should not be an issue.
+        // unwrap ok because it is not safe to continue running if the mutex is poisoned.
+        let mut r = self.recoverable_txes.lock().unwrap();
+        r.push(*tx);
+    }
+}
+
+#[async_trait]
+impl<'a, C: 'a> WriteAheadLog<'a, C> for DBWriteAheadLog<C>
+where
+    C: Serialize + DeserializeOwned + std::marker::Send + std::marker::Sync,
+{
+    type Guard = DBTxGuard<'a, C>;
+
+    #[must_use]
+    async fn begin_tx(
+        &'a self,
+        tx: &TransactionDigest,
+        cert: &C,
+    ) -> SuiResult<Option<DBTxGuard<'a, C>>> {
+        let _mutex_guard = self.mutex_table.acquire_lock(tx).await;
+
+        if self.log.contains_key(tx)? {
+            return Ok(None);
```

Comment:
I have trouble with the semantics of this `Ok(None)`. This really means: do not worry we are already processing this exact transaction. Should we retry later? (prbably not as the effects should be there).

Should we block until the transaction is processed? And only then return None?

Reply:
I don't have a good answer to this yet. I will leave a TODO for now - I may indeed want to have this block until the concurrent tx is finished. I think it will become clearer when i write the code that this is using